addons/product/models/product_template.py

```diff
@@ -938,6 +939,77 @@ def _get_first_possible_combination(self, parent_combination=None, necessary_val
         """
         return next(self._get_possible_combinations(parent_combination, necessary_values), self.env['product.template.attribute.value'])
 
+    def _cartesian_product(self, product_template_attribute_values, parent_combination):
+        """
+        Generate all possible combination for attributes values (aka cartesian product).
+        It is equivalent to itertools.product except it skips invalid partial combinations before they are complete.
+
+        Imagine the cartesian product of 'A', 'CD' and range(1_000_000) and let's say that 'A' and 'C' are incompatible.
+        If you use itertools.product or any normal cartesian product, you'll need to filter out of the final result
+        the 1_000_000 combinations that start with 'A' and 'C' . Instead, This implementation will test if 'A' and 'C' are
+        compatible before even considering range(1_000_000), skip it and and continue with combinations that start
+        with 'A' and 'D'.
+
+        It's necessary for performance reason because filtering out invalid combinations from standard Cartesian product
+        can be extremely slow
+
+        :param product_template_attribute_values: the values we want all the possibles combinations of.
+        One list of values by attribute line
+        :return: a generator of product template values combinations (in the form of lists)
+        """
+        if not product_template_attribute_values:
+            raise StopIteration
+
+        all_exclusions = self._get_own_attribute_exclusions()
+        # 0 means the value is acceptable, greater than 0 means it's rejected
+        current_exclusions = defaultdict(int)
+        for exclusion in self._get_parent_attribute_exclusions(parent_combination):
+            current_exclusions[exclusion] += 1
+        included_values_ids = set()
+
+        # -1 means no value has been picked for the attribute in the current (partial) combination
+        values_idx = [-1] * len(product_template_attribute_values)
```

Comment:
```python
# The following list reflects product_template_attribute_values_list
# For each attribute, instead of a list of values, it contains the index of the selected value
# -1 means no value has been picked for the attribute in the current (partial) combination
```

Reply:
And each index targets the list it is used on.